File: ignasr-app/src/test/java/limdongjin/ignasr/util/MultiPartUtilTests.java
package limdongjin.ignasr.util;

import limdongjin.ignasr.MyTestUtil;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.core.io.ClassPathResource;
import org.springframework.web.reactive.function.server.ServerRequest;
import reactor.core.publisher.Mono;

import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.UnsupportedAudioFileException;
import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.IOException;
import java.util.UUID;
import java.util.function.Function;

class MultiPartUtilTests {
    @Test
    void canLoadWavBlobFromServerRequest() throws IOException, UnsupportedAudioFileException {
        // Given
        var reqId = UUID.randomUUID().toString();
        var userId = UUID.randomUUID().toString();
        ClassPathResource classPathResource = new ClassPathResource("data/foo.wav");
        File file = classPathResource.getFile();
        AudioInputStream expectedAudioInputStream = AudioSystem.getAudioInputStream(file);
        String label = "ldj";

        // Prepare ServerRequest
        ServerRequest serverRequest = MyTestUtil.prepareServerRequest(classPathResource, reqId, label, userId);

        // Extract file, name from ServerRequest, Load Audio
        Function<String, Mono<byte[]>> fieldNameToBytesMono = MultiPartUtil.toFunctionThatFieldNameToBytesMono(serverRequest);
        Mono<byte[]> fileMono = fieldNameToBytesMono.apply("file");
        Mono<String> nameMono = fieldNameToBytesMono.apply("name").map(String::new);

        AudioInputStream actualAudioInputStream = AudioSystem.getAudioInputStream(new ByteArrayInputStream(fileMono.block()));

        // Verify
        Assertions.assertEquals(reqId, nameMono.block());
        Assertions.assertEquals(expectedAudioInputStream.getFormat().toString(), actualAudioInputStream.getFormat().toString());
        Assertions.assertEquals(expectedAudioInputStream.getFrameLength(), actualAudioInputStream.getFrameLength());

        System.out.println(actualAudioInputStream.getFormat());
        // example:
        // PCM_SIGNED 16000.0 Hz, 16 bit, mono, 2 bytes/frame, little-endian

        System.out.println(nameMono.block());
    }
}

File: ignasr-app-kt/src/test/kotlin/limdongjin/ignasr/util/MultiPartUtilTests.kt
package limdongjin.ignasr.util

import limdongjin.ignasr.MyTestUtil.prepareServerRequest
import limdongjin.ignasr.util.MultiPartUtil.toFunctionThatFieldNameToBytesMono
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.core.io.ClassPathResource
import java.io.ByteArrayInputStream
import java.io.IOException
import java.util.*
import javax.sound.sampled.AudioSystem
import javax.sound.sampled.UnsupportedAudioFileException


class MultiPartUtilTests {
    @Test
    @Throws(IOException::class, UnsupportedAudioFileException::class)
    fun canLoadWavBlobFromServerRequest() {
        // Given
        val reqId = UUID.randomUUID().toString()
        val userId = UUID.randomUUID().toString()
        val classPathResource = ClassPathResource("data/foo.wav")
        val file = classPathResource.file
        val expectedAudioInputStream = AudioSystem.getAudioInputStream(file)
        val label = "ldj"

        // Prepare ServerRequest
        val serverRequest = prepareServerRequest(classPathResource, reqId, label, userId)

        // Extract file, name from ServerRequest, Load Audio
        val fieldNameToBytesMono = toFunctionThatFieldNameToBytesMono(
            serverRequest!!
        )
        val fileMono = fieldNameToBytesMono.apply("file")
        val nameMono = fieldNameToBytesMono.apply("name").map { bytes: ByteArray? -> String(bytes!!) }
        val actualAudioInputStream = AudioSystem.getAudioInputStream(ByteArrayInputStream(fileMono.block()))

        // Verify
        Assertions.assertEquals(reqId, nameMono.block())
        Assertions.assertEquals(expectedAudioInputStream.format.toString(), actualAudioInputStream.format.toString())
        Assertions.assertEquals(expectedAudioInputStream.frameLength, actualAudioInputStream.frameLength)
        println(actualAudioInputStream.format)
        // example:
        // PCM_SIGNED 16000.0 Hz, 16 bit, mono, 2 bytes/frame, little-endian
        println(nameMono.block())
    }
}